components/region_cache_memory_engine/src/write_batch.rs

```diff
@@ -77,26 +100,89 @@ impl RangeCacheWriteBatch {
         let (entries_to_write, engine) = self.engine.handle_pending_range_in_loading_buffer(
             seq,
             std::mem::take(&mut self.pending_range_in_loading_buffer),
         );
         let guard = &epoch::pin();
-        entries_to_write
+        let entry_counts = entries_to_write.len() + self.buffer.len();
+        // Some entries whose ranges may be marked as evicted above, but it does not
+        // matter, they will be deleted later.
+        let res = entries_to_write
             .into_iter()
             .chain(std::mem::take(&mut self.buffer))
-            .try_for_each(|e| e.write_to_memory(&engine, seq, guard))
+            .try_for_each(|e| {
+                e.write_to_memory(seq, &engine, self.memory_controller.clone(), guard)
+            });
+        self.memory_controller.on_node_written(entry_counts);
+
+        if let Err(e) = self
+            .engine
+            .bg_worker_manager()
+            .schedule_task(BackgroundTask::DeleteRange(ranges_to_delete))
+        {
+            error!(
+                "schedule delete range failed";
+                "err" => ?e,
+            );
+            assert!(tikv_util::thread_group::is_shutdown(!cfg!(test)));
+        }
+
+        res
+    }
+
+    // return ranges that can be deleted from engine now
+    fn handle_ranges_to_evict(&mut self) -> Vec<CacheRange> {
+        if self.ranges_to_evict.is_empty() {
+            return vec![];
+        }
+        let mut core = self.engine.core.write();
+        let mut ranges = vec![];
+        let range_manager = core.mut_range_manager();
+        for r in std::mem::take(&mut self.ranges_to_evict) {
+            if range_manager.contains_range(&r) && range_manager.evict_range(&r) {
+                ranges.push(r);
+                continue;
+            }
+
+            if let Some((range, _, canceled)) = range_manager
+                .pending_ranges_loading_data
+                .iter_mut()
+                .find(|(range, ..)| range.contains_range(&r))
+            {
+                range_manager.ranges_being_deleted.insert(range.clone());
+                *canceled = true;
+            }
+        }
+        ranges
     }
 
     #[inline]
     pub fn set_range_cache_status(&mut self, range_cache_status: RangeCacheStatus) {
         self.range_cache_status = range_cache_status;
     }
 
-    fn process_cf_operation<F>(&mut self, entry: F)
+    fn process_cf_operation<F1, F2>(&mut self, entry_size: F1, entry: F2)
     where
-        F: FnOnce() -> RangeCacheWriteBatchEntry,
+        F1: FnOnce() -> usize,
+        F2: FnOnce() -> RangeCacheWriteBatchEntry,
     {
+        if !matches!(
+            self.range_cache_status,
+            RangeCacheStatus::Cached | RangeCacheStatus::Loading
+        ) || self.memory_usage_reach_hard_limit
+        {
+            return;
+        }
+
+        let memory_expect = entry_size();
+        if !self.memory_acquire(memory_expect) {
```

Comment:
When memory_acquire return false, should process_cf_operation return false as well and propagate it to the caller side? And the caller should immediately mark the region as not readable.
Otherwise, how can the caller know if the write to in-memory engine actually succeed or not?

Reply:
If the acquire failed, the ranges will be pushed into `ranges_to_evict`, and in `write_impl` theses ranges will be evicted. For the duration between them, the range is still readable.